app/vmalert/rule.go

```diff
@@ -268,3 +295,44 @@ func newTimeSeries(value float64, labels map[string]string, timestamp time.Time)
 	}
 	return ts
 }
+
+// Restore restores the state of active alerts basing on previously written timeseries.
+// Restore restores only Start field. Field State will be always Pending and supposed
+// to be updated on next Eval, as well as Value field.
+func (r *Rule) Restore(ctx context.Context, q datasource.Querier, lookback time.Duration) error {
+	// get the last datapoint in range
+	expr := fmt.Sprintf("last_over_time(%s{alertname=%q}[%ds])",
```

Comment:
This makes `vmalert` incompatible with Prometheus, which doesn't support `last_over_time` function. Probably, it would be better querying `/api/v1/query` with range query like `%s{alertname=%q}[%ds]`. It will return raw samples for the given time series over the given time range. Then you can get the last sample. See https://www.robustperception.io/extracting-raw-samples-from-prometheus for details.

Reply:
MetricsQL used intentionally since `vmalert` persists its state via remote-write protocol which isn't supported by Prometheus. So it is assumed that VM will be used as remote storage for `vmalert`. Put this explanation into code.